File: src/main/java/me/syari/ss/discord/internal/handle/GuildSetupNode.java
package me.syari.ss.discord.internal.handle;

import gnu.trove.iterator.TLongIterator;
import gnu.trove.map.TLongObjectMap;
import gnu.trove.map.hash.TLongObjectHashMap;
import gnu.trove.set.TLongSet;
import gnu.trove.set.hash.TLongHashSet;
import me.syari.ss.discord.api.utils.data.DataArray;
import me.syari.ss.discord.api.utils.data.DataObject;
import me.syari.ss.discord.internal.JDA;
import me.syari.ss.discord.internal.entities.Guild;
import org.jetbrains.annotations.NotNull;

import java.util.LinkedList;
import java.util.List;

public class GuildSetupNode {
    private final List<DataObject> cachedEvents = new LinkedList<>();
    private TLongObjectMap<DataObject> members;
    private TLongSet removedMembers;
    private DataObject partialGuild;
    private int expectedMemberCount = 1;
    boolean requestedChunk;
    boolean markedUnavailable = false;
    GuildSetupController.Status status = GuildSetupController.Status.INIT;
    private final long id;
    private final GuildSetupController controller;
    private final JDA api;

    GuildSetupNode(long id, @NotNull GuildSetupController controller) {
        this.id = id;
        this.controller = controller;
        api = controller.getJDA();
    }

    @Override
    public String toString() {
        return "GuildSetupNode[" + id + "|" + status + ']' +
                '{' +
                "expectedMemberCount=" + expectedMemberCount + ", " +
                "requestedChunk=" + requestedChunk + ", " +
                "markedUnavailable=" + markedUnavailable +
                '}';
    }

    @Override
    public int hashCode() {
        return Long.hashCode(id);
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof GuildSetupNode)) return false;
        GuildSetupNode node = (GuildSetupNode) obj;
        return node.id == id;
    }

    void updateStatus(GuildSetupController.Status status) {
        if (status != this.status) this.status = status;
    }

    void handleCreate(DataObject obj) {
        if (partialGuild == null) {
            partialGuild = obj;
        } else {
            for (String key : obj.keys()) {
                partialGuild.put(key, obj.opt(key).orElse(null));
            }
        }
        boolean unavailable = partialGuild.getBoolean("unavailable");
        this.markedUnavailable = unavailable;
        if (unavailable) return;
        ensureMembers();
    }

    boolean handleMemberChunk(DataArray arr) {
        if (partialGuild == null) return true;
        for (int index = 0; index < arr.length(); index++) {
            DataObject obj = arr.getObject(index);
            long id = obj.getObject("user").getLong("id");
            members.put(id, obj);
        }
        if (expectedMemberCount <= members.size() || !api.chunkGuild(id)) {
            completeSetup();
            return false;
        }
        return true;
    }

    void cacheEvent(@NotNull DataObject event) {
        cachedEvents.add(event);
        int cacheSize = cachedEvents.size();
        if (2000 <= cacheSize && cacheSize % 1000 == 0 && status == GuildSetupController.Status.CHUNKING) {
            controller.sendChunkRequest(id);
        }
    }

    private void completeSetup() {
        updateStatus(GuildSetupController.Status.BUILDING);
        TLongIterator iterator = removedMembers.iterator();
        while (iterator.hasNext()) {
            members.remove(iterator.next());
        }
        removedMembers.clear();
        Guild guild = api.getEntityBuilder().createGuild(id, partialGuild, expectedMemberCount);
        if (requestedChunk) {
            controller.ready(id);
        } else {
            controller.remove(id);
        }
        updateStatus(GuildSetupController.Status.READY);
        api.getClient().handle(cachedEvents);
        api.getEventCache().playbackCache(EventCache.Type.GUILD, id);
        guild.acknowledgeMembers();
    }

    private void ensureMembers() {
        expectedMemberCount = partialGuild.getInt("member_count");
        members = new TLongObjectHashMap<>(expectedMemberCount);
        removedMembers = new TLongHashSet();
        DataArray memberArray = partialGuild.getArray("members");
        if (!api.chunkGuild(id)) {
            handleMemberChunk(memberArray);
        } else if (memberArray.length() < expectedMemberCount && !requestedChunk) {
            updateStatus(GuildSetupController.Status.CHUNKING);
            controller.addGuildForChunking(id);
            requestedChunk = true;
        } else if (handleMemberChunk(memberArray) && !requestedChunk) {
            members.clear();
            updateStatus(GuildSetupController.Status.CHUNKING);
            controller.addGuildForChunking(id);
            requestedChunk = true;
        }
    }
}

File: src/main/java/me/syari/ss/discord/internal/handle/GuildSetupNode.kt
package me.syari.ss.discord.internal.handle

import gnu.trove.map.hash.TLongObjectHashMap
import gnu.trove.set.TLongSet
import gnu.trove.set.hash.TLongHashSet
import me.syari.ss.discord.api.data.DataArray
import me.syari.ss.discord.api.data.DataContainer
import me.syari.ss.discord.internal.entities.EntityBuilder
import me.syari.ss.discord.internal.requests.WebSocketClient
import java.util.LinkedList

class GuildSetupNode(private val id: Long) {
    private val cachedEvents: MutableList<DataContainer> = LinkedList()
    private var members = TLongObjectHashMap<DataContainer>()
    private var removedMembers: TLongSet? = null
    private var partialGuild: DataContainer? = null
    private var expectedMemberCount = 1
    private var requestedChunk = false
    private var status = GuildSetupController.Status.INIT

    override fun toString(): String {
        return "GuildSetupNode[$id|$status]{expectedMemberCount=$expectedMemberCount, requestedChunk=$requestedChunk}"
    }

    override fun hashCode(): Int {
        return java.lang.Long.hashCode(id)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is GuildSetupNode) return false
        return other.id == id
    }

    private fun updateStatus(status: GuildSetupController.Status) {
        if (status !== this.status) this.status = status
    }

    fun handleCreate(dataObject: DataContainer) {
        val notNulPartialGuild = partialGuild?.apply {
            for (key in dataObject.keys) {
                put(key, dataObject.get(key))
            }
        } ?: {
            partialGuild = dataObject
            dataObject
        }.invoke()
        val unavailable = notNulPartialGuild.getBoolean("unavailable") ?: false
        if (unavailable) return
        expectedMemberCount = notNulPartialGuild.getIntOrThrow("member_count")
        members = TLongObjectHashMap(expectedMemberCount)
        removedMembers = TLongHashSet()
        val memberArray = notNulPartialGuild.getArrayOrThrow("members")
        if (memberArray.size < expectedMemberCount && !requestedChunk) {
            updateStatus(GuildSetupController.Status.CHUNKING)
            GuildSetupController.addGuildForChunking(id)
            requestedChunk = true
        } else if (handleMemberChunk(memberArray) && !requestedChunk) {
            members.clear()
            updateStatus(GuildSetupController.Status.CHUNKING)
            GuildSetupController.addGuildForChunking(id)
            requestedChunk = true
        }
    }

    private fun handleMemberChunk(arr: DataArray): Boolean {
        if (partialGuild == null) return true
        for (index in 0 until arr.size) {
            val obj = arr.getContainerOrThrow(index)
            val id = obj.getContainerOrThrow("user").getLongOrThrow("id")
            members.put(id, obj)
        }
        if (expectedMemberCount <= members.size()) {
            completeSetup()
            return false
        }
        return true
    }

    fun cacheEvent(event: DataContainer) {
        cachedEvents.add(event)
        val cacheSize = cachedEvents.size
        if (2000 <= cacheSize && cacheSize % 1000 == 0 && status === GuildSetupController.Status.CHUNKING) {
            GuildSetupController.sendChunkRequest(id)
        }
    }

    private fun completeSetup() {
        updateStatus(GuildSetupController.Status.BUILDING)
        removedMembers?.let { removedMembers ->
            val iterator = removedMembers.iterator()
            while (iterator.hasNext()) {
                members.remove(iterator.next())
            }
            removedMembers.clear()
        }
        partialGuild?.let { EntityBuilder.createGuild(id, it) }
        if (requestedChunk) {
            GuildSetupController.ready(id)
        } else {
            GuildSetupController.remove(id)
        }
        updateStatus(GuildSetupController.Status.READY)
        WebSocketClient.handle(cachedEvents)
        EventCache.playbackCache(EventCache.Type.GUILD, id)
    }
}